ouroboros-consensus/src/Ouroboros/Consensus/Protocol/PBFT.hs

```diff
@@ -214,26 +222,101 @@ instance (PBftCrypto c, Typeable c) => OuroborosTag (PBft c) where
-              gkSigners = Seq.length (Seq.filter (== gk) signers)
+          let totalSigners = chainStateSize chainState
+              gkSigners = maybe 0 Seq.length $ Map.lookup gk chainState
           when (totalSigners >= winSize && gkSigners > wt)
             $ throwError (PBftExceededSignThreshold totalSigners gkSigners)
-          return $! takeR (winSize + 2*k) chainState Seq.|> (gk, blockSlot b)
+          return $! insertSigner gk (blockSlot b) $ pruneChainState (winSize + 2*k) chainState
     where
       PBftParams{..} = pbftParams
       PBftFields{..} = headerPBftFields cfg b
       winSize = fromIntegral pbftSignatureWindow
       SecurityParam (fromIntegral -> k) = pbftSecurityParam
       wt = floor $ pbftSignatureThreshold * fromIntegral winSize
-      -- Take the rightmost n elements of a sequence
-      takeR :: Integral i => i -> Seq a -> Seq a
-      takeR (fromIntegral -> n) s = Seq.drop (Seq.length s - n - 1) s
 
   rewindChainState _ cs mSlot = case mSlot of
-    Origin  -> Just Seq.empty
-    At slot -> case Seq.takeWhileL (\(_, s) -> s <= slot) cs of
-        _ Seq.:<| _ -> Just cs
-        _           -> Nothing
+    Origin  -> Just Map.empty
+    At slot -> let oldCs = Map.map (Seq.takeWhileL (\s -> s <= slot)) cs in
+      if (any (not . Seq.null) $ Map.elems oldCs) then Just oldCs else Nothing
+
+-- | Prune the chain state to the given size by dropping the signers in the
+-- oldest slots.
+--
+-- Examples
+--
+-- $setup
+-- >>> :set -XOverloadedLists
+-- >>> :set -XScopedTypeVariables
+-- >>> import Test.QuickCheck
+-- >>> :{
+--   newtype TestChainState = TestChainState (Map Int (Seq Int))
+--   instance Arbitrary TestChainState where
+--     arbitrary = do
+--       numElts <- arbitrary
+--       ub <- arbitrary
+--       return . TestChainState $ Map.fromList
+--         [(idx, elts) | idx <- [0..numElts]
+--                      , let elts = Seq.fromList [i | i <- [0..ub], i `mod` numElts == idx]
+--         ]
+-- :}
+
+-- prop> \cs -> pruneChainState (chainStateSize cs) cs == cs
+--
+-- prop> \n (TestChainState cs) -> n > 0 && (chainStateSize cs) > n ==> (chainStateSize (pruneChainState n cs)) == n
+--
+-- >>> :{
+--   let (cs :: Map Char (Seq Int)) = [('a', [1,2,5]), ('b', [3, 4, 6])]
+--   in pruneChainState 2 cs
+-- :}
+-- fromList [('a',fromList [5]),('b',fromList [6])]
+--
+pruneChainState :: forall k v. (Show k, Show v, Ord k, Ord v) => Int -> Map k (Seq v) -> Map k (Seq v)
+pruneChainState toSize cs = go
+    cs
+    (sortOn snd . catMaybes $ strengthr . (fmap (Seq.lookup 0)) <$> (Map.toAscList cs))
+    (max 0 $ chainStateSize cs - toSize)
+  where
+    go :: Map k (Seq v)
+       -> [(k, v)]
+       -> Int
+       -> Map k (Seq v)
+    go fromCS idx toDrop = if toDrop <= 0 then fromCS else case idx of
+      [] -> fromCS
+      (gk,_):xs@((_,nextLowest):_) ->
+        let (newSeq, numDropped) = fromJust $ dropWhileL (< nextLowest) <$> Map.lookup gk fromCS
```

Comment:
I see some `fromJust`s, can't we avoid those? I'm going to try to fully understand this function :slightly_smiling_face: and see if it is possible.

Reply:
I update the index and map separately, in order to not have to rebuild the index each time. So the compiler can't verify that, if something exists in the index, it also exists in the map. Almost certainly this could be rewritten to avoid it, but I'm not convinced it's worth it.

Reply:
I think the following implementation is easier to understand and more "functional", but I'm not sure about efficiency. Maybe we can use it as the spec for the real implementation and compare both in the property tests?

```haskell
pruneChainState toSize cs = unspreadVs
                          . take toSize
                            -- Sort on @v@, from large to small
                          . sortOn (Down . fst)
                          . spreadVs
                          $ cs
  where
    spreadVs :: Map k (Seq v) -> [(v, k)]
    spreadVs kvs = [(v, k) | (k, vs) <- Map.toAscList kvs, v <- toList vs]

    -- @vs@ from large to small
    unspreadVs :: [(v, k)] -> Map k (Seq v)
    unspreadVs = foldl' addVK (Seq.empty <$ cs)
      -- Start with all the keys from the initial map, but with empty 'Seq's.

    -- @Seq v@ is from small to large, the given @v@ is smaller than the @vs@
    -- in @Seq v@.
    addVK :: Map k (Seq v) -> (v, k) -> Map k (Seq v)
    addVK m (v, k) = Map.adjust (v Seq.<|) k m
```
Complexity:
```
spread: O(v)
sort: O(v log v)
take: O(toSize)
unspread: O(v log k)
= O(v log v) as k is small 
```
We should actually benchmark the two implementations, because I would prefer using this, as long as it is more efficient.

Reply:
I think this will be quite a lot less efficient. The implementation I gave should be something like `O(toDrop * k log k)` (It could be made into `O(k log k + k * toDrop)`, but this didn't seem worth it). Crucially, `toDrop` and `k` are both likely to be small (the modal value of `toDrop` should be 1 for Byron), whereas `v` will be large (4k or similar, where k is now the security param)

Reply:
Indeed, microbenchmarks show that your version is much faster.